src/compiler/checker.ts

```diff
@@ -8016,6 +8154,65 @@ module ts {
             }
         }
 
+        function getAwaitedType(type: Type, fallbackType?: Type, seen: boolean[] = []): Type {
+            var seen: boolean[] = [];
+            if (!type || !globalIPromiseType || !(type.flags & TypeFlags.ObjectType)) {
+                return fallbackType;
+            }
+            seen[type.id] = true;
+            if (checkTypeRelatedTo(type, globalIPromiseType, assignableRelation, undefined)) {
+                var thenProp = getPropertyOfType(type, "then");
+                var thenType = getTypeOfSymbol(thenProp);
+                var thenSignatures = getSignaturesOfType(thenType, SignatureKind.Call);
+                var onFulfilledParameterType = getUnionType(map(thenSignatures, signature => getTypeAtPosition(getErasedSignature(signature), 0)));
+                var onFulfilledParameterSignatures = getSignaturesOfType(onFulfilledParameterType, SignatureKind.Call);
+                var awaitedType = getUnionType(map(onFulfilledParameterSignatures, signature => getTypeAtPosition(getErasedSignature(signature), 0)));
+                if (seen[awaitedType.id]) return fallbackType;
+                return getAwaitedType(awaitedType, /*fallbackType*/ awaitedType, seen);
+            } else {
+                if (isTypeAssignableTo(type, thenableType)) {
+                    error(null, ts.Diagnostics.Type_for_await_does_not_have_a_valid_callable_then_member);
+                }
+                return fallbackType;
+            }
+        }
+
+        function checkAwaitableReturnType(node: SignatureDeclaration, returnType: Type): Type {
+            // an async function has a valid return type if the type has a construct signature that takes
+            // in an `initializer` function that in turn supplies a `resolve` function as one of its arguments
+            // and results in an object with a callable `then` signature.
+            if (globalIPromiseConstructorType) {
+                if (!returnType) {
+                    returnType = getReturnTypeOfSignature(getSignatureFromDeclaration(node));
+                }
+
+                if (returnType && returnType.symbol) {
+                    var links = getSymbolLinks(returnType.symbol);
+                    if (!links.promiseType) {
+                        var type = getTypeOfSymbol(returnType.symbol);

```

Comment:
That’s a bit tricky but not impossible. Instead I’m getting the symbol for the return type and getting the type of that symbol. I check the return type as an expression later on.

Also, I’ve changed this code to:

```TypeScript
var declaredType = returnType.symbol ? getTypeOfSymbol(returnType.symbol) : emptyObjectType;
```

If I wanted to look up the name of the symbol, I’d have to do some more complex logic, especially if the name of the type has no local name in the current symbol table (such as if the type reference has a qualified name). I can do this, but I’m letting checkExpressionOrQualifiedName and checkCollisionWithAwaiterVariablesInGeneratedCode do this work.

From: Jason Freeman [mailto:notifications@github.com]
Sent: Friday, February 13, 2015 5:30 PM
To: Microsoft/TypeScript
Cc: Ron Buckton
Subject: Re: [TypeScript] Async Functions (#1781)


In src/compiler/checker.ts<https://github.com/Microsoft/TypeScript/pull/1781#discussion_r24710241>:

> +            }

> +        }

> +

> +        function checkAwaitableReturnType(node: SignatureDeclaration, returnType: Type): Type {

> +            // an async function has a valid return type if the type has a construct signature that takes

> +            // in an `initializer` function that in turn supplies a `resolve` function as one of its arguments

> +            // and results in an object with a callable `then` signature.

> +            if (globalIPromiseConstructorType) {

> +                if (!returnType) {

> +                    returnType = getReturnTypeOfSignature(getSignatureFromDeclaration(node));

> +                }

> +

> +                if (returnType && returnType.symbol) {

> +                    var links = getSymbolLinks(returnType.symbol);

> +                    if (!links.promiseType) {

> +                        var type = getTypeOfSymbol(returnType.symbol);

I don't understand what this is doing? Why would type be different from returnType? I thought you have to take the name of returnType, and then look it up in the value space.

—
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/TypeScript/pull/1781/files#r24710241>.